### PR TITLE
Che bato/play from other zone

### PR DESCRIPTION
**Pull request by CheBato**
Added the functionality to play from other locations beside from hand.

- Added Ezra bridge

EzraBridger

- Added Brutal traditions (https://swudb.com/card/shd/038/brutal-traditions)

BrutalTraditions

### CODE REVIEW

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts
Comment (adbronico) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:3]: Also a test for when Ezra's controller's deck is empty

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (adbronico) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:3]: Can we add a test that ensures the Brutal Tradition can be played from discard multiple times per phase?

File: server/game/cards/01_SOR/units/EzraBridger.ts
Comment (adbronico) [inline server/game/cards/01_SOR/units/EzraBridger.ts:30]: I think we can consider removing optional: true here. The card does say "may" but in this case Leave it on top of the deck serves the same purpose.

File: server/game/core/Player.js
Comment (adbronico) [inline server/game/core/Player.js:72]: @AMMayberry1 Do we want two records for one play type like this, or do we want to enhance the location to accept an array?

**Review by adbronico — DISMISSED** [review]
There's a few more test cases we should add. I'll let Veld review the AbilityContext changes since I'm not as confident in what should be changed there if anything.

File: server/game/cards/01_SOR/units/EzraBridger.ts
Comment (CheBato) [reply server/game/cards/01_SOR/units/EzraBridger.ts:30]: Yes Sir i'll remove the extra optional

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts
Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:48]: yes sir i updated the test cases

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts
Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:37]: Added the TODO

Comment (CheBato) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:37]: Added a TODO to the lookAtSystem!

File: server/game/gameSystems/PutIntoPlaySystem.ts
Comment (AMMayberry1) [inline server/game/gameSystems/PutIntoPlaySystem.ts:62]: Maybe we just remove the facedown check completely since I think it's actually impossible for it to fail now? The only facedown locations are resource and deck and those now have exceptions

File: server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts
Comment (AMMayberry1) [inline server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts:15]: Let's move this file and the related changes out of this PR since it's for the Kylo PR and therefore isn't being tested. If we check it in then people will assume it's tested and working.

File: test/server/cards/01_SOR/units/EzraBridger.spec.ts
Comment (AMMayberry1) [reply test/server/cards/01_SOR/units/EzraBridger.spec.ts:26]: I've been encouraging people to do them for now, it makes the long multi-case tests cleaner. We're going to have do a massive refactor anyway if we ever actually do get to the point of having a better test harness, and I don't think having or not having these will have a real impact on the amount of work.

If anything they would make it easier to refactor since we can tell right away which lines are just state reset and aren't part of a specific case.

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:38]: Actually we can test two things at once here - if we merge this into the above test but do it after `moveToNextActionPhase()`, we confirm that the condition is over when the phase changes and that the card is not playable when the condition is not met

File: server/game/core/ability/PlayCardAction.ts
Comment (CheBato) [reply server/game/core/ability/PlayCardAction.ts:62]: Yes sir my i apoligize.

Comment (CheBato) [reply server/game/core/ability/PlayCardAction.ts:62]: i think the , was leftover from when i attempted to add playType here to test it.

File: server/game/core/Player.js
Comment (CheBato) [reply server/game/core/Player.js:686]: ooooh i didn't know you can do this thats pretty neet!

File: server/game/core/gameSteps/prompts/ResourcePrompt.ts
Comment (CheBato) [reply server/game/core/gameSteps/prompts/ResourcePrompt.ts:70]: yessir i will go back to space being there

File: server/game/gameSystems/PutIntoPlaySystem.ts
Comment (CheBato) [reply server/game/gameSystems/PutIntoPlaySystem.ts:62]: removed the facedown check!

Comment (CheBato) [reply server/game/gameSystems/PutIntoPlaySystem.ts:62]: Unfortunatelly by removing the check multiple tests failed like timely intervention so I'm assuming we need to keep it.

File: server/game/gameSystems/GameSystemLibrary.ts
Comment (CheBato) [reply server/game/gameSystems/GameSystemLibrary.ts:209]: Set it to default seems like it works :)

File: server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts
Comment (CheBato) [reply server/game/stateWatchers/CardsDiscardedThisPhaseWatcher.ts:15]: Yes sir

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (CheBato) [reply test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:3]: added the test!

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (CheBato) [reply test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts]: yessir! Made them into one!

File: server/game/cards/01_SOR/units/EzraBridger.ts
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:27]: I've had a realization recently that cards of the form "look at the top card of your deck. do [X] with it" might actually be best represented with `ifYouDo` instead of `then` with a `thenCondition`. The card doesn't explicitly say "if you do" but mechanically it's understood that if you don't look at the top card for any reason (including the deck being empty) then the second part of the effect should fizzle.

Could you try making that change here? There is an option to have an `ifYouDoContext` but I think you won't need it since you're just reading context.source. If you do want to use it, you might need to merge main since I just did a bug fix for it

File: server/game/cards/01_SOR/units/EzraBridger.ts
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:33]: I think there needs to be a parenthesis on the noAction like `AbilityHelper.immediateEffects.noAction()`. If you hit an issue where that makes this button disappear, then try `immediateEffects.noAction({ hasLegalTarget: true })`

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:28]: This check isn't really needed

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:31]: Any reason to do it this way instead of starting it in discard? Could just use like atst instead of vigilant honor guards to kill wampa

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:40]: Can remove these checks, we have other tests confirming that attached upgrades give stats correctly. Just checking the upgrade is attached is enough

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:46]: Don't need this check, we know Vanquish works

File: test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts
Comment (AMMayberry1) [inline test/server/cards/02_SHD/upgrades/BrutalTraditions.spec.ts:28]: One last check we can do I realized - now that we have your feature for highlighting selectable cards in the action phase, can you add a check that it correctly appears in the highlight list in the action phase? Same for other places we do this check in here

File: server/game/cards/01_SOR/units/EzraBridger.ts
Comment (AMMayberry1) [inline server/game/cards/01_SOR/units/EzraBridger.ts:23]: Do we actually need this context? The top card of deck is accessible via the regular context, so we could just do:

```typescript
ifYouDo: {
    title: 'You may play it, discard it, or leave it on top of your deck.',
    targetResolver: {
        mode: TargetMode.Select,
        choices: {
            ['Play it']: AbilityHelper.immediateEffects.playCardFromOutOfPlay((context) => ({ target: context.source.controller.getTopCardOfDeck() }))),
            ['Discard it']: AbilityHelper.immediateEffects.discardSpecificCard((context) => ({ target: context.source.controller.getTopCardOfDeck() }))),
            ['Leave it on top of your deck']: AbilityHelper.immediateEffects.noAction({ hasLegalTarget: true }),
        }
    }
}
```
If you do decide you prefer it the way it is, let's rename `thenContext` -> `ifYouDoContext`

Comment (CheBato) [reply server/game/cards/01_SOR/units/EzraBridger.ts:23]: you are right it is much cleaner this way. I changed it.

**Review by AMMayberry1 — APPROVED** [review]

**Pull request merged**
